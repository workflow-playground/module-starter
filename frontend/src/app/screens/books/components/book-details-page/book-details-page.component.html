<ds-book-details [book]="(book$ | async)!">
  <button
    (click)="onClick()"
    class="w-8 h-8 flex items-center justify-center rounded shadow"
    [ngClass]="{
      'bg-primary-p300-main': (isSelectedBookInCollection$ | async) === false,
      'bg-red-700': (isSelectedBookInCollection$ | async)
    }"
  >
    <ds-icon [size]="20" [name]="icons.Save" [color]="iconColors.white"></ds-icon>
  </button>
</ds-book-details>
