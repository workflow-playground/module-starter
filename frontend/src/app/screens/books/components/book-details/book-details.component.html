<div class="w-full relative divide-gray-200 rounded-lg bg-white text-center shadow p-4">
  <div class="flex mb-5">
    <div class="mr-5 h-full" *ngIf="thumbnail">
      <ds-image [width]="150" [src]="thumbnail"></ds-image>
    </div>

    <div class="text-left">
      <h1 class="text-2xl text-neutral-n500 leading-tight">{{ title }}</h1>
      <h2 class="text-xl text-neutral-n300 p-0 mb-2 leading-tight" *ngIf="subtitle">{{ subtitle }}</h2>

      <div class="flex items-center mb-2" *ngIf="authors.length > 0">
        <span class="text-neutral-n300 mr-2">{{ 'BOOKS.DETAILS.AUTHORS' | translate }}</span>
        <span
          *ngFor="let author of authors"
          class="inline-flex items-center rounded-full bg-primary-p300-main px-2 py-1 m-1 text-xs font-medium text-primary-p100 ring-1 ring-inset ring-primary-p300"
          >{{ author }}</span
        >
      </div>

      <div class="flex items-center mb-2" *ngIf="published">
        <span class="text-neutral-n300 mr-2">{{ 'BOOKS.DETAILS.PUBLISHED' | translate }}</span>
        <span class="text-neutral-n500">{{ published }}</span>
      </div>

      <div class="flex items-center mb-2" *ngIf="language">
        <span class="text-neutral-n300 mr-2">{{ 'BOOKS.DETAILS.LANGUAGE' | translate }}</span>
        <span class="text-neutral-n500">{{ language }}</span>
      </div>
    </div>
  </div>
  <blockquote class="text-left italic text-neutral-n500" [innerHtml]="description"></blockquote>
  <div class="absolute top-4 right-4">
    <ng-content></ng-content>
  </div>
</div>
