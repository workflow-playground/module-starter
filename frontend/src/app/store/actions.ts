import { actions as books } from './slices/books.store';
import { actions as collection } from './slices/collection.store';
import { actions as search } from './slices/search.store';
import { actions as auth } from './slices/auth.store';
import { actions as layout } from './slices/layout.store';
import { actions as loginPage } from './slices/login-page.store';
import { actions as user } from './slices/user.store';

export { books, collection, search, auth, loginPage, layout, user };
