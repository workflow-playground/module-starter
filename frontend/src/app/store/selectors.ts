import { selectors as books } from './slices/books.store';
import { selectors as collection } from './slices/collection.store';
import { selectors as search } from './slices/search.store';
import { selectors as layout } from './slices/layout.store';
import { selectors as auth } from './slices/auth.store';
import { selectors as loginPage } from './slices/login-page.store';
import { selectors as router } from './slices/router.store';

export {
  router,
  books,
  collection,
  search,
  layout,
  auth,
  loginPage
}
