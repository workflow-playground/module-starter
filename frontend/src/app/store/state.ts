import { RouterReducerState } from '@ngrx/router-store';

import * as books from './slices/books.store';
import * as collection from './slices/collection.store';
import * as search from './slices/search.store';
import * as auth from './slices/auth.store';
import * as loginPage from './slices/login-page.store';
import * as layout from './slices/layout.store';

/**
 * As mentioned, we treat each reducer like a table in a database. This means
 * our top level state interface is just a map of keys to inner state types.
 */
export interface State {
  [books.featureKey]: books.State;
  [collection.featureKey]: collection.State;
  [search.featureKey]: search.State;
  [auth.featureKey]: auth.State;
  [loginPage.featureKey]: loginPage.State;
  [layout.featureKey]: layout.State;
  router: RouterReducerState;
}
