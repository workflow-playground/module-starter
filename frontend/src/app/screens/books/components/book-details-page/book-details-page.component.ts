import { ChangeDetectionStrategy, Component } from '@angular/core';
import { IconColor, IconName } from '@ds24/elements';

import { Store } from '@ngrx/store';
import { Observable, combineLatest, take } from 'rxjs';

import * as actions from '../../../../store/actions';
import * as selectors from '../../../../store/selectors';
import { Book } from '../../../../types/book.types';

@Component({
  selector: 'ds-book-details-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  templateUrl: './book-details-page.component.html',
})
export class BookDetailsPageComponent {
  icons: typeof IconName = IconName;
  iconColors: typeof IconColor = IconColor;

  book$ = this.store.select(selectors.books.selectSelectedBook) as Observable<Book>;
  isSelectedBookInCollection$ = this.store.select(selectors.collection.isSelectedBookInCollection);

  constructor(private store: Store) {}

  onClick() {
    combineLatest([this.book$, this.isSelectedBookInCollection$])
      .pipe(take(1))
      .subscribe(([book, isSelectedBookInCollection]) => {
        if (!isSelectedBookInCollection) {
          this.store.dispatch(actions.books.addBook({ book }));
        } else {
          this.store.dispatch(actions.books.removeBook({ book }));
        }
      });
  }
}
