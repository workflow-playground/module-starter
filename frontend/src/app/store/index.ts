import * as selectors from './selectors';
import * as actions from './actions';
import { State } from './state';

export {
  selectors,
  State,
  actions
}
