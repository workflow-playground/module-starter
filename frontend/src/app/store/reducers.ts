import { isDevMode } from '@angular/core';
import { MetaReducer, ActionReducerMap } from '@ngrx/store';
import { routerReducer } from '@ngrx/router-store';

import { State } from './state';

/**
 * Every reducer module's default export is the reducer function itself. In
 * addition, each module should export a type or interface that describes
 * the state of the reducer plus any selector functions. The `* as`
 * notation packages up all of the exports into a single object.
 */

import * as books from './slices/books.store';
import * as collection from './slices/collection.store';
import * as search from './slices/search.store';
import * as auth from './slices/auth.store';
import * as loginPage from './slices/login-page.store';
import * as layout from './slices/layout.store';

/**
 * Our state is composed of a map of action reducer functions.
 * These reducer functions are called with each dispatched action
 * and the current or initial state and return a new immutable state.
 */
export const rootReducers: ActionReducerMap<State> = {
  [books.featureKey]: books.reducer,
  [collection.featureKey]: collection.reducer,
  [search.featureKey]: search.reducer,
  [auth.featureKey]: auth.reducer,
  [loginPage.featureKey]: loginPage.reducer,
  [layout.featureKey]: layout.reducer,
  router: routerReducer,
};

/**
 * By default, @ngrx/store uses combineReducers with the reducer map to compose
 * the root meta-reducer. To add more meta-reducers, provide an array of meta-reducers
 * that will be composed to form the root meta-reducer.
 */
export const metaReducers: MetaReducer<State>[] = isDevMode() ? [] : [];
