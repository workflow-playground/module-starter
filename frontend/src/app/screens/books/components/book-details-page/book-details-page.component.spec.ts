import { ComponentFixture, TestBed } from '@angular/core/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { IconModule } from '@ds24/elements';
import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { MemoizedSelector } from '@ngrx/store';

import { BookDetailsPageComponent } from './book-details-page.component';
import { BookDetailsComponent } from '../book-details/book-details.component';

import { State, actions, selectors } from '../../../../store';
import { bookMock } from '../../../../types/book.types.mock';

describe('Book Details Page', () => {
  let fixture: ComponentFixture<BookDetailsPageComponent>;
  let mockStore: MockStore<State>;
  let instance: BookDetailsPageComponent;
  let selectSelectedBook: MemoizedSelector<State, any>;
  let isSelectedBookInCollection: MemoizedSelector<State, boolean>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [NoopAnimationsModule, IconModule],
      declarations: [BookDetailsPageComponent, BookDetailsComponent],
      providers: [provideMockStore()],
    });

    fixture = TestBed.createComponent(BookDetailsPageComponent);
    instance = fixture.componentInstance;
    mockStore = TestBed.inject(MockStore);

    selectSelectedBook = mockStore.overrideSelector(selectors.books.selectSelectedBook, null);
    isSelectedBookInCollection = mockStore.overrideSelector(selectors.collection.isSelectedBookInCollection, false);

    jest.spyOn(mockStore, 'dispatch');
  });

  it('should compile', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a books.addBook action when onClick is called and the book is not in the collection', () => {
    selectSelectedBook.setResult(bookMock);
    isSelectedBookInCollection.setResult(false);

    const action = actions.books.addBook({ book: bookMock });

    instance.onClick();

    expect(mockStore.dispatch).toHaveBeenLastCalledWith(action);
  });

  it('should dispatch a books.removeBook action on removeFromCollection', () => {
    selectSelectedBook.setResult(bookMock);
    isSelectedBookInCollection.setResult(true);

    const action = actions.books.removeBook({ book: bookMock });

    instance.onClick();

    expect(mockStore.dispatch).toHaveBeenLastCalledWith(action);
  });
});
