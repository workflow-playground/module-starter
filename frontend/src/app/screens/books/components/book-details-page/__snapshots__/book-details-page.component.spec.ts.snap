// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book Details Page should compile 1`] = `
<ds-book-details-page
  book$={[Function _Store]}
  iconColors={[Function Object]}
  icons={[Function Object]}
  isSelectedBookInCollection$={[Function _Store]}
  store={[Function _MockStore]}
>
  <ds-book-details>
    <div
      class="w-full relative divide-gray-200 rounded-lg bg-white text-center shadow p-4"
    >
      <div
        class="flex mb-5"
      >
        <div
          class="text-left"
        >
          <h1
            class="text-2xl text-neutral-n500 leading-tight"
          >
          </h1>
        </div>
      </div>
      <blockquote
        class="text-left italic text-neutral-n500"
      />
      <div
        class="absolute top-4 right-4"
      >
        <button
          class="w-8 h-8 flex items-center justify-center rounded shadow bg-primary-p300-main"
        >
          <ds-icon>
            <save-icon>
              <svg
                fill="none"
                height="20px"
                style="height: 20px; width: 20px;"
                viewBox="0 0 24 24"
                width="20px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  class="fill-white"
                  clip-rule="evenodd"
                  d="M17.3846 2H2V22H22V6.76923L17.3846 2ZM14.7846 3.53846V8.15385H7.44615V3.53846H14.7846ZM20.4615 20.4615H3.53846V3.53846H5.90769V9.69231H16.3231V3.53846H16.7231L20.4615 7.38462V20.4615ZM15.8462 14.8999H5.78467V16.4384H15.8462V14.8999ZM5.78467 17.2693H15.8462V18.8077H5.78467V17.2693Z"
                  fill-rule="evenodd"
                />
              </svg>
            </save-icon>
          </ds-icon>
        </button>
      </div>
    </div>
  </ds-book-details>
</ds-book-details-page>
`;
